app/src/org/commcare/android/util/StorageUtils.java

```diff
@@ -85,6 +86,32 @@ public int compare(Integer lhs, Integer rhs) {
         for(int i = 0 ; i < ids.size() ; ++i) {
             records[i] = storage.read(ids.elementAt(i).intValue());
         }
-        return records;
+        return filterByCurrentApp(records);
     }
+
+    /**
+     * Filter the given list of forms down to only those that are part of the currently seated app
+     */
+    public static FormRecord[] filterByCurrentApp(FormRecord[] forms) {
```

Comment:
Instead of iterating over the input array twice and setting/checking elements for nullity, I would recommend doing one pass through 'forms' and add relevant forms to a result Vector that you then turn into an array and return.